tests/test_runner.py

```diff
@@ -0,0 +1,262 @@
+    runner,
+    step_execute_mock,
+    step_validate_results_mock,
+    step_is_configured_mock,
+):
+    with pytest.raises(ValidateRequirementsFailure) as excinfo:
+        runner.validate_all_requirements()
+
+    assert all(isinstance(exc, PrerequisiteFailed) for exc in excinfo.value.exceptions)
+
+    step_is_configured_mock.assert_not_called()
+    step_execute_mock.assert_not_called()
+    step_validate_results_mock.assert_not_called()
+
+
+@pytest.fixture()
+def expected_successful_run_result(runner_step, expected_step_config):
+    return StepExecutionResult(
+        step=runner_step,
+        is_configured=False,
+        is_enabled=True,
+        has_run=True,
+        run_exception=None,
+        config_model=expected_step_config,
+    )
+
+
+def test_execute_step_returns_correct_result(
+    runner,
+    step_execute_mock,
+    expected_step_config,
+    expected_successful_run_result,
+    runner_step,
+    step_is_configured_mock,
+    step_validate_results_mock,
+):
+    result = runner._execute_step(runner_step)
+
+    assert result == expected_successful_run_result
+
+    step_is_configured_mock.assert_called_once_with(expected_step_config)
+    step_execute_mock.assert_called_once_with(expected_step_config)
+    step_validate_results_mock.assert_not_called()
+
+
+def test_execute_all_returns_correct_results(
+    runner,
+    step_execute_mock,
+    expected_step_config,
+    expected_successful_run_result,
+    step_is_configured_mock,
+    step_validate_results_mock,
+):
+    results = runner.execute_all()
+
+    assert results == [expected_successful_run_result]
+
+    step_is_configured_mock.assert_called_once_with(expected_step_config)
+    step_execute_mock.assert_called_once_with(expected_step_config)
+    step_validate_results_mock.assert_not_called()
+
+
+def test_exception_during_execute_step_is_included_in_result(
+    runner,
+    step_execute_mock,
+    runner_step,
+    expected_step_config,
+    step_is_configured_mock,
+    step_validate_results_mock,
+    expected_successful_run_result,
+):
+    step_execute_mock.side_effect = Exception()
+    result = runner._execute_step(runner_step)
+
+    assert result == dataclasses.replace(
+        expected_successful_run_result, run_exception=step_execute_mock.side_effect
+    )
+
+    step_is_configured_mock.assert_called_once_with(expected_step_config)
+    step_execute_mock.assert_called_once_with(expected_step_config)
+    step_validate_results_mock.assert_not_called()
+
+
+def test_exception_during_execute_all_is_included_in_result(
+    runner,
+    step_execute_mock,
+    runner_step,
+    expected_step_config,
+    step_is_configured_mock,
+    step_validate_results_mock,
+    expected_successful_run_result,
+):
+    step_execute_mock.side_effect = Exception()
+    result = runner.execute_all()
+
+    assert result == [
+        dataclasses.replace(
+            expected_successful_run_result, run_exception=step_execute_mock.side_effect
+        )
```

Comment:
Maybe you could instantiate a new `StepExecutionResult` instead of replacing something on the `succesful` fixture